web/modules/weather_routes/src/Controller/LocationAndGridRouteController.php

```diff
@@ -113,28 +113,80 @@ public function redirectFromGrid($wfo, $gridX, $gridY)
+        );
+    }
+
+    public function servePlacePage($state, $place)
     {
         try {
-            $location = $this->dataLayer->databaseFetch(
-                "SELECT
-                    ST_X(point) AS lon,
-                    ST_Y(point) AS lat
-                FROM weathergov_geo_places
-                WHERE
-                    state LIKE :state
-                    AND
-                    name LIKE :place",
-                [":state" => $state, ":place" => $place],
-            );
-
-            if ($location !== false) {
-                $url = Url::fromRoute("weather_routes.point", [
-                    "lat" => $location->lat,
-                    "lon" => $location->lon,
+            $known = $this->getKnownPlace($state, $place);
+
+            // If we don't know this place, immediately return a 404. There's no
+            // point doing any further processing.
+            if ($known == false) {
+                throw new NotFoundHttpException();
+            }
+
+            // If the place name has a space in it, we'll need to redirect the
+            // user to the canonical URL.
+            $redirectForSpaces = str_contains($place, " ");
+
+            // Unescape the place name so we can compare it to the place name
+            // we got from the database.
+            $place = str_replace("_", " ", $place);
+            $place = str_replace(",", "/", $place);
+
+            // If the place name we got has a space in it, or if the state or
+            // place do not strictly match what's in our database, we'll send
+            // the user a redirect to our canonical place URL.
+            if (
+                $redirectForSpaces ||
+                $state != $known->state ||
+                $place != $known->name
+            ) {
+                // We'll send them back to this same route, but we'll use
+                // the state and place names from our database, with the
+                // place name escaped to handle spaces and slashes.
+                // (There are two places in American Samoa with slashes in
+                // their names, but slashes are URL path characters so we
+                // don't want to persist them!)
+                $url = Url::fromRoute("weather_routes.place", [
+                    "state" => $known->state,
+                    "place" => str_replace(
+                        "/",
+                        ",",
+                        str_replace(" ", "_", $known->name),
+                    ),
                 ]);
                 return new RedirectResponse($url->toString());
             }
+
+            // If we're here, we know the location and the requested URL is
+            // already canonical, so continue with the rendering process.
+            // If anyone figures out how to pass the lat/lon from here on
+            // through the process, that'd be dandy – then the template
+            // variable preprocessor thingy wouldn't have to redo this query
```

Comment:
Thinking off the top of my head: it might be possible now to do away with the preprocess hook altogether for the point page route, and instead use the "theme object" structure to define what kinds of data the template for the point page expects (it has a confusing name, it's more like a schema for the expected assoc array a given template or node would want to render). You can see [these examples for the AFD page](https://github.com/weather-gov/weather.gov/blob/main/web/modules/weather_routes/weather_routes.module#L5) where that happens. I believe all AFD data fetching occurs now in the route handler. We could consider doing the same for our location pages.
  
If you think that makes sense, I'll write up a ticket for the work and take it from there.

Reply:
That seems dope. That variable prerender hook thing is so ugly, and it is completely undiscoverable. Future devs will be thankful if the code is in a sensible place like a route handler, where it totally makes sense.

Reply:
I've documented in a separate ticket here https://github.com/weather-gov/weather.gov/issues/2078